pkg/jwt/facade.go

```diff
@@ -1,26 +1,34 @@
 package jwt
 
 import (
-	"crypto/rsa"
-
 	"gopkg.in/square/go-jose.v2"
 	"gopkg.in/square/go-jose.v2/jwt"
 )
 
-// CreateRS256SignerAndValidator creates a token signer and validator pair with the given properties from the
-// specified RSA private key.
-func CreateRS256SignerAndValidator(issuerID string, audience jwt.Audience, key *rsa.PrivateKey, keyID string) (jose.Signer, Validator, error) {
-	keyStore := NewSingleKeyStore(key.Public(), keyID)
-	validator := NewRS256Validator(keyStore, issuerID, audience)
+// SignerFactory produces jose.Signer on demand.
+type SignerFactory struct {
+	keyStore PrivateKeyGetter
+	keyID    string
+}
+
+// CreateSigner creates jose.Signer based on underlying private key.
+func (f *SignerFactory) CreateSigner() (jose.Signer, error) {
 	signingKey := jose.SigningKey{
 		Algorithm: jose.RS256,
-		Key:       key,
+		Key:       f.keyStore.Key(f.keyID),
 	}
-	signer, err := jose.NewSigner(signingKey, new(jose.SignerOptions).WithType("JWT").WithHeader("kid", keyID))
-	if err != nil {
-		return nil, nil, err
+	return jose.NewSigner(signingKey, new(jose.SignerOptions).WithType("JWT").WithHeader("kid", f.keyID))
+}
+
+// CreateRS256SignerAndValidator creates a token signer and validator pair with the given properties from the
+// specified RSA private key.
+func CreateRS256SignerAndValidator(issuerID string, audience jwt.Audience, privateKeyStore PrivateKeyGetter, publicKeyStore PublicKeyGetter, keyID string) (*SignerFactory, Validator) {
+	validator := NewRS256Validator(publicKeyStore, issuerID, audience)
+	signerFactory := &SignerFactory{
+		keyStore: privateKeyStore,
+		keyID:    keyID,
 	}
-	return signer, validator, nil
+	return signerFactory, validator
 }
```

Comment:
This `CreateRS256SignerAndValidator` function takes two sets of parameters, which are used to construct two objects.
The parameters for the returned objects constructors do not overlap, so I'd suggest splitting this function into `CreateSignerFactory` (sic!) and `CreateRS256Validator` with the corresponding parameters. And then we would realize that one doesn't need another function to create a Validator.